src/pudl/io_managers.py

```diff
@@ -695,36 +695,85 @@ def filter_for_freshest_data(
         Each row in our SQLite database includes all the facts for one context/filing
         pair.
 
-        If one context is represented in multiple filings, we take the facts from the
-        most recently-published filing.
+        If one context is represented in multiple filings, we take the most
+        recently-reported non-null value.
 
-        This means that if a recently-published filing does not include a value for a
-        fact that was previously reported, then that value will remain null. We do not
-        forward-fill facts on a fact-by-fact basis.
+        This means that if a utility reports a non-null value, then later
+        either reports a null value for it or simply omits it from the report,
+        we keep the old non-null value, which may be erroneous. This appears to
+        be fairly rare, affecting < 0.005% of reported values.
         """
+
+        def __apply_diffs(
+            duped_groups: pd.core.groupby.DataFrameGroupBy,
+        ) -> pd.DataFrame:
+            """Take the latest reported non-null value for each group."""
+            return duped_groups.last()
+
+        def __best_snapshot(
+            duped_groups: pd.core.groupby.DataFrameGroupBy,
+        ) -> pd.DataFrame:
+            """Take the row that has most non-null values out of each group."""
+            return duped_groups.apply(
+                lambda df: df.assign(count=df.count(axis="columns"))
+                .sort_values(by="count", ascending=True)
+                .tail(1)
+            ).drop(columns="count")
+
+        def __compare_dedupe_methodologies(
+            apply_diffs: pd.DataFrame, best_snapshot: pd.DataFrame
+        ):
+            """Compare deduplication methodologies.
+
+            By cross-referencing these we can make sure that the apply-diff
```

Comment:
😓

Reply:
hm yeah it is surprisingly annoying!! I'm now semi regretting suggesting it but i think this looks okay. i like that it makes it clear that there are a few options and 

This suggestion is probably oos for this pr but i wonder if this type of check that assesses methodological choices we've made in validation tests when it gets a lil too complicated. I've personally struggled with this distinction and the tension around testing with a lot of extra code that is not necessary for the actually thing you want to consistently produce. but i don't think this is tooo much!